### PR TITLE
ARROW-11837: [C++][Dataset] expose originating Fragment on ScanTask

### PR DESCRIPTION
**Pull request by lidavidm**
Now scan tasks keep their originating fragment alive, removing the need for applications to track the mapping themselves.

### CODE REVIEW

File: cpp/src/arrow/dataset/dataset.h
Comment (lidavidm) [inline cpp/src/arrow/dataset/dataset.h:46]: A further improvement would be to also expose `std::string Fragment::name()` or similar for logging purposes; file fragments would return the filename, while other fragments would return some fixed identifier (much like type_name).

Comment (bkietz) [reply cpp/src/arrow/dataset/dataset.h:46]: That sounds good to me, but I think it should be called `ToString()`

File: cpp/src/arrow/dataset/file_base.cc
Comment (bkietz) [inline cpp/src/arrow/dataset/file_base.cc:330]: :+1:

File: cpp/src/arrow/dataset/scanner.cc
Comment (bkietz) [inline cpp/src/arrow/dataset/scanner.cc:80]:
```suggestion
  auto fragment = std::make_shared<InMemoryFragment>(std::move(schema), std::move(batches));
```

**Review by bkietz — CHANGES_REQUESTED** [review]
GLib is broken by the modification to ScanTask construction, but other than that this looks good

**Comment by lidavidm** [comment]
Thanks. Looks like c_glib is going to take me a bit as they don't have bindings for Fragment and while I could ad-hoc fix that one location, if they intend to mirror the Datasets API as best they can, then we should actually introduce the Fragment binding.

**Comment by lidavidm** [comment]
Let's hope these GLib bindings I whipped up just now actually work! We probably want another set of eyes for those…

**Comment by kou** [comment]
I'll review this.

**Comment by lidavidm** [comment]
Ah, this is broken until I fully define GADInMemoryFragment...let me fix that up

**Comment by lidavidm** [comment]
Looks like things finally pass after fixing the GLib and Ruby bindings. I opted not to bind Fragment::Scan since the API there is being reworked at the moment anyways.

**Review by kou — APPROVED** [review]
+1

I've reviewed GLib and Ruby parts. They almost looks good to me. Some minor fixes are directly pushed.

I think that we can merge this once CI is green.

**Comment by lidavidm** [comment]
Thank you for the review and for patching up my very poor bindings :sweat_smile:

**Comment by kou** [comment]
CI is green. I'll merge this.

Thanks for updating GLib and Ruby parts too!

**Pull request closed without merging**
